Refactor part two

karczex opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmemkv-python/20)
<!-- Reviewable:end -->

Commit:
Add test for repetitive remove the same element

Commit:
Throw KeyError if key not exist in dictionary

For compatibility with dictionary api

lukaszstolarczuk reviewed (commented):
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 1 of 6 files reviewed, 10 unresolved discussions (waiting on @karczex)

___
*[pmemkv/kvengine.cc, line 50 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwE6qjSEF4LRufvmvE0:-LwE6qjSEF4LRufvmvE1:biu4foh) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L50)):*
> ```cpp
> 	const char *value;
> 	Py_ssize_t length;
> }PmemkvValueBufferObject;
> ```

add space after `}`
___
*[pmemkv/kvengine.cc, line 59 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEAfAO0Nu2F85TcHPT:-LwEAfAO0Nu2F85TcHPU:b-hg5zts) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L59)):*
> ```cpp
> };
> 
> static int PmemkvValueBufferObject_getbuffer(PyObject *obj,Py_buffer *view, int flags)
> ```

1. add space before `Py_buffer`
2. are flags used anywhere?
___
*[pmemkv/kvengine.cc, line 65 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEFwH-0eBZv5CFFv9N:-LwEFwH-0eBZv5CFFv9O:b-yec5d8) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L65)):*
> ```cpp
> 		return -1;
> 	}
> 	PmemkvValueBufferObject* self =  (PmemkvValueBufferObject*)obj;
> ```

one redundant space before `(Pmem...`
___
*[pmemkv/kvengine.cc, line 66 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEJIZD3y1p3g4bLoFY:-LwEJIZD3y1p3g4bLoFZ:b-rft3t0) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L66)):*
> ```cpp
> 	}
> 	PmemkvValueBufferObject* self =  (PmemkvValueBufferObject*)obj;
> 	view->obj = (PyObject*)self;
> ```

why not just assign here function's param `obj`?
___
*[pmemkv/kvengine.cc, line 96 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEcfM67nJsRhIlo-PP:-LwEcfM67nJsRhIlo-PQ:b-xft0mp) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L96)):*
> ```cpp
> PmemkvValueBuffer_init(PmemkvValueBufferObject *self)
> {
> 	if(self !=NULL) {
> ```

pls add space after `!=`
___
*[pmemkv/kvengine.cc, line 128 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEZssB9CehYdLvnfE6:-LwEZssB9CehYdLvnfE7:bwumoy3) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L128)):*
> ```cpp
> 	.tp_setattro = 0,
> 	.tp_as_buffer = &PmemkvValueBuffer_as_buffer,
> 	.tp_flags =  Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
> ```

redundant one space after `=`

here and some below
___
*[pmemkv/kvengine.cc, line 179 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEassk0lvbA9pSuU4L:-LwEassk0lvbA9pSuU4M:b-xft0mp) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L179)):*
> ```cpp
> static int Pmemkv_init(PmemkvObject *self)
> {
> 	if(self !=NULL) {
> ```

pls add space after `!=`
___
*[pmemkv/kvengine.cc, line 266 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEfBtL9D6UgmknX7Nq:-LwEfBtM9p3sqSrMRqXJ:b-x6qj89) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L266)):*
> ```cpp
> 	PyObject *args = PyTuple_New(2);
> 	if (PyTuple_SetItem(args, 0, key_obj) != 0)
> 		return 0;
> ```

you don't want to return something else...?

here and the `if` below
___
*[pmemkv/pmemkv.py, line 60 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEj6Tv2QGBHPMAVnyj:-LwEj6Tv2QGBHPMAVnyk:b1fpli5) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/pmemkv.py#L60)):*
> ```Python
> 
>     def __getitem__(self, key):
>         if key not  in self:
> ```

one redundant space after `not`
___
*[tests/pmemkv_tests.py, line 258 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEmKwd8j_l_OaQM7Vg:-LwEmKwe1XZuzD_K4npr:bnue0ri) ([raw file](https://github.com/pmem/pmemkv-python/blob/8cfddc3fcbf77bb881729757d2fb9552921ed63e/tests/pmemkv_tests.py#L258)):*
> ```Python
>         self.assertEqual(self.key, r"<1>,<2>,")
> 
>         db.put(r"记!", r"RR")
> ```

why removing these?


<!-- Sent from Reviewable.io -->

igchor reviewed (commented):
Please run reset-author on commits

> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20#-:-LwEvb8q6AAoibv3S-zz:bnly94c)* status: 1 of 6 files reviewed, 12 unresolved discussions (waiting on @karczex and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 266 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEfBtL9D6UgmknX7Nq:-LwI56s5AzVHfgN7LLQA:b-5yy7bg) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L266)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

you don't want to return something else...?

here and the `if` below
</blockquote></details>

Right, it should be an error probably
___
*[pmemkv/kvengine.cc, line 234 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI40do3ZdiGMTkJe4u:-LwI40do3ZdiGMTkJe4v:bnony54) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L234)):*
> ```cpp
> value_callback(const char* value, size_t valuebyte, void* context) {
> 	PmemkvValueBufferObject *entry = PyObject_New(PmemkvValueBufferObject, &PmemkvValueBufferType);
> 	if(entry) {
> ```

what if this is NULL?
___
*[pmemkv/kvengine.cc, line 269 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI6pPd9BnbCB5Na7NR:-LwI6pPd9BnbCB5Na7NS:b-kw9oph) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L269)):*
> ```cpp
> 	if (PyTuple_SetItem(args, 1, (PyObject *)entry) != 0)
> 		return 0;
> 	PyObject *res = PyObject_CallObject((PyObject *) context, args);
> ```

I think we should check if an exception was throw and if it was return some non-zero value from this function (this will stop iterating on pmemkv). This should also be documented that if any exception escapes the python callback it will stop iteration.


<!-- Sent from Reviewable.io -->

karczex reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 1 of 6 files reviewed, 12 unresolved discussions (waiting on @karczex and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 50 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwE6qjSEF4LRufvmvE0:-LwILUEIC4FxG8luOfcW:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L50)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

add space after `}`
</blockquote></details>

Done.
___
*[pmemkv/kvengine.cc, line 65 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEFwH-0eBZv5CFFv9N:-LwIMSP25iGoLdW9SS5k:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L65)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

one redundant space before `(Pmem...`
</blockquote></details>

Done.
___
*[pmemkv/kvengine.cc, line 128 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEZssB9CehYdLvnfE6:-LwIN3wf5i9omRyW4JhT:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L128)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

redundant one space after `=`

here and some below
</blockquote></details>

Done.
___
*[pmemkv/pmemkv.py, line 60 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEj6Tv2QGBHPMAVnyj:-LwIW5zJA4irMuoKiG_e:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/pmemkv.py#L60)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

one redundant space after `not`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

igchor reviewed (commented):
> Reviewed 2 of 6 files at r1, 2 of 2 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 3 of 6 files reviewed, 14 unresolved discussions (waiting on @igchor, @karczex, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 285 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI_Rg3D4KIX9Pfh0WR:-LwI_Rg3D4KIX9Pfh0WS:buqy5wa) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L285)):*
> ```cpp
> 	if(PyErr_Occurred() != NULL)
> 		return NULL;
> 	if (result != PMEMKV_STATUS_OK) {
> ```

After you implement the feature with stopping iteration this result can be PMEMKV_STATUS_STOPPED_BY_CB. What we want to do in that case? Should we throw an exception in that case? (I would say yes).
___
*[pmemkv/kvengine.cc, line 286 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI_JbGCnUsyp5xhCQl:-LwI_JbGCnUsyp5xhCQm:bmp7zfm) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L286)):*
> ```cpp
> 		return NULL;
> 	if (result != PMEMKV_STATUS_OK) {
> 		PyErr_SetObject(PyExc_Exception, PyLong_FromLong(result));
> ```

what about pmemkv_errormsg in exception message?


<!-- Sent from Reviewable.io -->

karczex commented:
___
*[tests/pmemkv_tests.py, line 258 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEmKwd8j_l_OaQM7Vg:-LwIjvkz3sfHwEiIY-KA:bfhxr6z) ([raw file](https://github.com/pmem/pmemkv-python/blob/8cfddc3fcbf77bb881729757d2fb9552921ed63e/tests/pmemkv_tests.py#L258)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

why removing these?
</blockquote></details>

Removing part of test related to removed API function get_keys_strings()


<!-- Sent from Reviewable.io -->

karczex commented:
___
*[pmemkv/kvengine.cc, line 286 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI_JbGCnUsyp5xhCQl:-LwIk_8mDOOqKUWQ5wSH:b-zayz3o) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L286)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

what about pmemkv_errormsg in exception message?
</blockquote></details>

Here should be thrown PmemkvException (which should be added) instead of general exception.  In such exception, error message should be added.  It's related to this  issue:  https://github.com/pmem/pmemkv-python/issues/18 

  


<!-- Sent from Reviewable.io -->

igchor reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 3 of 7 files reviewed, 15 unresolved discussions (waiting on @igchor, @karczex, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 50 at r5](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwrMu650tJlbUI1xiZ0:-LwrMu650tJlbUI1xiZ1:b7vcgh4) ([raw file](https://github.com/pmem/pmemkv-python/blob/c42eaa3ca86ebde330d5c6d057830d45f4f801ba/pmemkv/kvengine.cc#L50)):*
> ```cpp
> 	const char *value;
> 	Py_ssize_t length;
> } PmemkvValueBufferObject;
> ```

What happens if user accesses memory which is outside of the specified buffer (for example, we give use a buffer which points to binary array of size 10 and the user tries to access element 11)? Ideally this should cause an exception.


<!-- Sent from Reviewable.io -->

Commit:
Setup exceptions

igchor reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 0 of 7 files reviewed, 16 unresolved discussions (waiting on @igchor, @karczex, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 255 at r8](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Lx_iFJA0s8Ni9LY3GNb:-Lx_iFJA0s8Ni9LY3GNc:bj172qe) ([raw file](https://github.com/pmem/pmemkv-python/blob/8adf42767cb5e1ebe22ffaf0b120d6cd12bccbfd/pmemkv/kvengine.cc#L255)):*
> ```cpp
> value_callback(const char* value, size_t valuebyte, void* context) {
> 	PmemkvValueBufferObject *entry = PyObject_New(PmemkvValueBufferObject, &PmemkvValueBufferType);
> 	Py_INCREF(entry);
> ```

Why? I thought that PyObject_New already gives object with refcount == 1


<!-- Sent from Reviewable.io -->

igchor reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 0 of 7 files reviewed, 16 unresolved discussions (waiting on @igchor, @karczex, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 50 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwE6qjSEF4LRufvmvE0:-Lx_ilVP1JEnXnPas2lV:bldu6wu) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L50)):*
<details><summary><i>Previously, karczex wrote…</i></summary><blockquote>

Done.
</blockquote></details>

we could use some clang-format for those...


<!-- Sent from Reviewable.io -->

Commit:
Add test for repetitive get the same element

Commit:
Add unit test for rising pmemkv exceptions

Commit:
Introduce Buffer Protocol

Non-copy access to variables and keys in callback functions passed to
pmemkv_NI_get, pmemkv_NI_get_all, pmemkv_NI_get_above, pmemkv_NI_get_below,
pmemkv_NI_get_between. Way to do it, is to create additional object type
PmemkvValueBufferObject, avialable only in callback functions. This object
implements Buffer Protocol. This idicate usage of memoryview inside callback
functions.
Function, which returns value as variable (through copy) was renamed
from get() to get_string(). get() currently gives access to data
throught buffer protocol.
As usage of very similar lambda functions in callbacks implemetation was error
prone, it was moved to named functions value_callback, key_callback and
key_value_callback.

Existing tests are adapted to all those changes and new tests for
callbacks are added

Commit:
Cleanup basic_example

Adjust example to changes in API, and make is little more
readable

Commit:
Move PmemkvObject fields initialization to init

Commit:
Remove string versions of functions

As BufferProtocol API was introduced, string versions of functions with
callback are redundant

Commit:
Properly handle empty key in __getitem__

Commit:
Change tests according to improved imports

Commit:
Fix tests for exception handling

Commit:
Change basic example according to improved imoprts

Commit:
Fix NaughtyStrings test

Commit:
Add test for running lambda expression in callback

Commit:
Add clang-format configuration

It's copied from pmemkv and SortIncludes: false was added.

In python c extensions order of includes matters, so it should not be
sorted automatically

more info: https://docs.python.org/3.8/c-api/intro.html#include-files

karczex reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 0 of 8 files reviewed, 16 unresolved discussions (waiting on @igchor, @karczex, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 50 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwE6qjSEF4LRufvmvE0:-Ly-1SpK1vxs5GW9ehY-:b-kfwepn) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L50)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

we could use some clang-format for those...
</blockquote></details>

I used clang format on all patches.
___
*[pmemkv/kvengine.cc, line 59 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEAfAO0Nu2F85TcHPT:-Ly-1gpu6T8Vulwrw-bk:b-kfwepn) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L59)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

1. add space before `Py_buffer`
2. are flags used anywhere?
</blockquote></details>

I used clang format on all patches.
___
*[pmemkv/kvengine.cc, line 96 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEcfM67nJsRhIlo-PP:-Ly-1jAX-6Eoh2gZzQxS:b-kfwepn) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L96)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

pls add space after `!=`
</blockquote></details>

I used clang format on all patches.
___
*[pmemkv/kvengine.cc, line 179 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEassk0lvbA9pSuU4L:-Ly-29YD0GAuSawpyw1n:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L179)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

pls add space after `!=`
</blockquote></details>

Done.
___
*[pmemkv/kvengine.cc, line 286 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI_JbGCnUsyp5xhCQl:-Ly-21Z60CJdbyyKHqW8:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L286)):*
<details><summary><i>Previously, karczex wrote…</i></summary><blockquote>

Here should be thrown PmemkvException (which should be added) instead of general exception.  In such exception, error message should be added.  It's related to this  issue:  https://github.com/pmem/pmemkv-python/issues/18 

  
</blockquote></details>

Done.
___
*[tests/pmemkv_tests.py, line 258 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEmKwd8j_l_OaQM7Vg:-Ly-2MD20ZxCn9h33i4u:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/8cfddc3fcbf77bb881729757d2fb9552921ed63e/tests/pmemkv_tests.py#L258)):*
<details><summary><i>Previously, karczex wrote…</i></summary><blockquote>

Removing part of test related to removed API function get_keys_strings()
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
Add test for accesses to memory outside of the specified buffer

karczex reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 0 of 8 files reviewed, 16 unresolved discussions (waiting on @igchor, @karczex, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 59 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEAfAO0Nu2F85TcHPT:-Ly-2Vvq4NpYojv-haFM:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L59)):*
<details><summary><i>Previously, karczex wrote…</i></summary><blockquote>

I used clang format on all patches.
</blockquote></details>

Done.
___
*[pmemkv/kvengine.cc, line 66 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEJIZD3y1p3g4bLoFY:-Ly-N2BU0444DtleeyrU:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L66)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

why not just assign here function's param `obj`?
</blockquote></details>

Done.
___
*[pmemkv/kvengine.cc, line 266 at r1](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwEfBtL9D6UgmknX7Nq:-Ly-IXCw0Ild0GVAudV2:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/53bef7bdec59a58d86e8ae68a9dd985b6dcc9859/pmemkv/kvengine.cc#L266)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Right, it should be an error probably
</blockquote></details>

Done.
___
*[pmemkv/kvengine.cc, line 234 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI40do3ZdiGMTkJe4u:-Ly-NAI0DINVX4Yn7nW5:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L234)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

what if this is NULL?
</blockquote></details>

Done.
___
*[pmemkv/kvengine.cc, line 269 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI6pPd9BnbCB5Na7NR:-Ly-JhUz0W_-NNj904WM:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L269)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

I think we should check if an exception was throw and if it was return some non-zero value from this function (this will stop iterating on pmemkv). This should also be documented that if any exception escapes the python callback it will stop iteration.
</blockquote></details>

Done.
___
*[pmemkv/kvengine.cc, line 285 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI_Rg3D4KIX9Pfh0WR:-Ly-JttHCJRwswUHKFIm:b-653k30) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L285)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

After you implement the feature with stopping iteration this result can be PMEMKV_STATUS_STOPPED_BY_CB. What we want to do in that case? Should we throw an exception in that case? (I would say yes).
</blockquote></details>

If callback function raise exception, this exception is forwarded. In case of other fail without exception if PMEMKV_STATUS_STOPPED_BY_CB is returned pmemkv_NI.StoppedByCallback is raised. 
___
*[pmemkv/kvengine.cc, line 50 at r5](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwrMu650tJlbUI1xiZ0:-Ly-6oa_C9RxjiWx3luA:b-531nbg) ([raw file](https://github.com/pmem/pmemkv-python/blob/c42eaa3ca86ebde330d5c6d057830d45f4f801ba/pmemkv/kvengine.cc#L50)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

What happens if user accesses memory which is outside of the specified buffer (for example, we give use a buffer which points to binary array of size 10 and the user tries to access element 11)? Ideally this should cause an exception.
</blockquote></details>

Test for such scenario added. Indeed, it causes exception.


<!-- Sent from Reviewable.io -->

karczex reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 0 of 8 files reviewed, 16 unresolved discussions (waiting on @igchor and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 255 at r8](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Lx_iFJA0s8Ni9LY3GNb:-Ly-bOgb0dfD4nO9VTkE:bctvfh1) ([raw file](https://github.com/pmem/pmemkv-python/blob/8adf42767cb5e1ebe22ffaf0b120d6cd12bccbfd/pmemkv/kvengine.cc#L255)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Why? I thought that PyObject_New already gives object with refcount == 1
</blockquote></details>

PyTuple_SetItem steals reference, so additional INCREF and DECREF in this function guarantee, that entry object wouldn't be garbage collected inside context, which is python callback function. 

Reference created by PyObject_New is removed by DECREF on args


<!-- Sent from Reviewable.io -->

llugin reviewed (commented):
> Reviewed 5 of 7 files at r8, 3 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: all files reviewed, 19 unresolved discussions (waiting on @igchor, @karczex, and @lukaszstolarczuk)

___
*[pmemkv/pmemkv.py, line 139 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly3lBcV-RDUNkARAq0d:-Ly3lBcV-RDUNkARAq0e:bre9uji) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/pmemkv/pmemkv.py#L139)):*
> ```Python
>     # Gets the value for the given key from pmemkv datastore.
>     # Takes key from the end user and returns the value.
>     def get(self, key, func):
> ```

1. not sure if part of this refactor but it's better if functions/classes documentations are docstrings, not comments. 

2a. if it is a public interface then the meaning of func should be explained. 2b. probably this could be an optional argument?
___
*[tests/pmemkv_tests.py, line 260 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly3p_2h5YejrES4122m:-Ly3p_2h5YejrES4122n:b-njw7pi) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/tests/pmemkv_tests.py#L260)):*
> ```Python
>             db = Database(self.engine, '{\"path\":1234, \"size\": 1073741824}')
>         with self.assertRaises(pmemkv.ConfigTypeError):
>             db = Database(self.engine, '{\"path\":1234, \"size\": 1073741824}')
> ```

raw string instead of escaping " could be more handy here
___
*[tests/pmemkv_tests.py, line 460 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly3qFFmAXZYBVezk0mM:-Ly3qFFmAXZYBVezk0mN:bavdkmy) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/tests/pmemkv_tests.py#L460)):*
> ```Python
>             def __init__(self):
>                 self.result = None
>             def call(self, key):
> ```

you could also try using special method \_\_call\_\_ here https://docs.python.org/3/reference/datamodel.html#object.__call__


<!-- Sent from Reviewable.io -->

igchor reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: all files reviewed, 18 unresolved discussions (waiting on @igchor, @karczex, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 263 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly9GeAf4YkeC1pMFMFc:-Ly9GeAf4YkeC1pMFMFd:b-jl86k9) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/pmemkv/kvengine.cc#L263)):*
> ```cpp
> 		PyObject_New(PmemkvValueBufferObject, &PmemkvValueBufferType);
> 	Py_INCREF(entry);
> 	if (entry) {
> ```

please throw some exception if null?


<!-- Sent from Reviewable.io -->

igchor reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: all files reviewed, 18 unresolved discussions (waiting on @karczex and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 269 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI6pPd9BnbCB5Na7NR:-Ly9VWdV2dUm_bmF9Eve:btage0k) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L269)):*
<details><summary><i>Previously, karczex wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Is this documented?
___
*[pmemkv/kvengine.cc, line 285 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI_Rg3D4KIX9Pfh0WR:-Ly9T-x3B-3zvODWpXL8:b-3el10q) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L285)):*
<details><summary><i>Previously, karczex wrote…</i></summary><blockquote>

If callback function raise exception, this exception is forwarded. In case of other fail without exception if PMEMKV_STATUS_STOPPED_BY_CB is returned pmemkv_NI.StoppedByCallback is raised. 
</blockquote></details>

Hm, shouldn't we rethrow the exception which was thrown inside of the callback? Or is this already happening in the if (PyErr_Occured())? Is there a test for that?
___
*[pmemkv/kvengine.cc, line 255 at r8](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Lx_iFJA0s8Ni9LY3GNb:-Ly9RcDV9lUhEnK4_u3_:b6qlfyo) ([raw file](https://github.com/pmem/pmemkv-python/blob/8adf42767cb5e1ebe22ffaf0b120d6cd12bccbfd/pmemkv/kvengine.cc#L255)):*
<details><summary><i>Previously, karczex wrote…</i></summary><blockquote>

PyTuple_SetItem steals reference, so additional INCREF and DECREF in this function guarantee, that entry object wouldn't be garbage collected inside context, which is python callback function. 

Reference created by PyObject_New is removed by DECREF on args
</blockquote></details>

as agreed, this can be removed + please add test for deleting the obj inside python callback
___
*[tests/pmemkv_tests.py, line 2 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly9Uq-o08LhQh5qcNJC:-Ly9Uq-pBFBeVIyabvBV:b-nljfzp) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/tests/pmemkv_tests.py#L2)):*
> ```Python
> '''
>  * Copyright 2019, Intel Corporation
> ```

please fix copyrights (-2020). ALso in other files.


<!-- Sent from Reviewable.io -->

karczex reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 6 of 8 files reviewed, 19 unresolved discussions (waiting on @karczex, @llugin, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 276 at r10](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LyABugI4a6osdn9WyRz:-LyABugI4a6osdn9WyS-:bebo6y6) ([raw file](https://github.com/pmem/pmemkv-python/blob/15b03c8b0522bdd075db1d60970c2311d66b56f8/pmemkv/kvengine.cc#L276)):*
> ```cpp
> 	} else {
> 		PyErr_SetObject(PyExc_Exception,
> 				PyLong_FromLong(PMEMKV_STATUS_STOPPED_BY_CB));
> ```

Need to update this according to exception dispatcher patch


<!-- Sent from Reviewable.io -->

Commit:
Add test for deleting PmemkvValueBufferObject inside python callback

karczex reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 4 of 7 files reviewed, 18 unresolved discussions (waiting on @igchor, @karczex, @llugin, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 285 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI_Rg3D4KIX9Pfh0WR:-LyDtgOe0Q9PWkZcelSs:b-ibjcod) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L285)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Hm, shouldn't we rethrow the exception which was thrown inside of the callback? Or is this already happening in the if (PyErr_Occured())? Is there a test for that?
</blockquote></details>

	if (PyErr_Occurred() != NULL)
		return NULL;
forward exception form callback
if exception was not raised in callback, but result from pmemkv_get_all is different than PMEMKV_STATUS_OK, throw exception according to status

there is test for exception raised by callback: test_get_exception_in_callback. 
___
*[pmemkv/kvengine.cc, line 255 at r8](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Lx_iFJA0s8Ni9LY3GNb:-LyDtChWC0KI-Y1YgAIN:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/8adf42767cb5e1ebe22ffaf0b120d6cd12bccbfd/pmemkv/kvengine.cc#L255)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

as agreed, this can be removed + please add test for deleting the obj inside python callback
</blockquote></details>

Done.
___
*[pmemkv/kvengine.cc, line 263 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly9GeAf4YkeC1pMFMFc:-LyDtDIW6T99eVDitoo-:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/pmemkv/kvengine.cc#L263)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

please throw some exception if null?
</blockquote></details>

Done.
___
*[tests/pmemkv_tests.py, line 2 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly9Uq-o08LhQh5qcNJC:-LyDzPt0DkNefZoPGd-V:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/tests/pmemkv_tests.py#L2)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

please fix copyrights (-2020). ALso in other files.
</blockquote></details>

Done.
___
*[tests/pmemkv_tests.py, line 260 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly3p_2h5YejrES4122m:-LyE-8vbBF-GS4YLpScz:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/tests/pmemkv_tests.py#L260)):*
<details><summary><i>Previously, llugin wrote…</i></summary><blockquote>

raw string instead of escaping " could be more handy here
</blockquote></details>

Done.
___
*[tests/pmemkv_tests.py, line 460 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly3qFFmAXZYBVezk0mM:-LyE-ALU7GU57Is4G8w_:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/tests/pmemkv_tests.py#L460)):*
<details><summary><i>Previously, llugin wrote…</i></summary><blockquote>

you could also try using special method \_\_call\_\_ here https://docs.python.org/3/reference/datamodel.html#object.__call__
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
Add brief documentation in docstring

To view documentation:
	import pmemkv
	help(pmemkv)
	help(pmemkv.pmemkv)

igchor reviewed (changes requested):
> Reviewed 1 of 6 files at r11.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 3 of 7 files reviewed, 15 unresolved discussions (waiting on @karczex, @llugin, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 275 at r11](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LyEfXZrEZpCrdpYp8VH:-LyEfXZrEZpCrdpYp8VI:buyjxoi) ([raw file](https://github.com/pmem/pmemkv-python/blob/302ff9774aea561b6d604f637bcc835dbc0af401/pmemkv/kvengine.cc#L275)):*
> ```cpp
> 		Py_XDECREF(args);
> 	} else {
> 		PyErr_SetString(ExceptionDispatcher[PMEMKV_STATUS_STOPPED_BY_CB].exception,
> ```

MemoryError


<!-- Sent from Reviewable.io -->

karczex reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20)* status: 3 of 7 files reviewed, 15 unresolved discussions (waiting on @igchor, @llugin, and @lukaszstolarczuk)

___
*[pmemkv/kvengine.cc, line 269 at r3](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LwI6pPd9BnbCB5Na7NR:-LyEhIn71IH7zqd0MhV9:b-ozjd95) ([raw file](https://github.com/pmem/pmemkv-python/blob/31325d72231b4f5bd6c36757475dab91130e86e7/pmemkv/kvengine.cc#L269)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Is this documented?
</blockquote></details>

Some brief documentation is added in this PR. Next step is to make it better.
___
*[pmemkv/kvengine.cc, line 275 at r11](https://reviewable.io/reviews/pmem/pmemkv-python/20#-LyEfXZrEZpCrdpYp8VH:-LyEhtd345uE0HMvADBB:b-896fix) ([raw file](https://github.com/pmem/pmemkv-python/blob/302ff9774aea561b6d604f637bcc835dbc0af401/pmemkv/kvengine.cc#L275)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

MemoryError
</blockquote></details>

Done.
___
*[pmemkv/pmemkv.py, line 139 at r9](https://reviewable.io/reviews/pmem/pmemkv-python/20#-Ly3lBcV-RDUNkARAq0d:-LyEi-e_-WRhPKJToRLP:b-d1p32r) ([raw file](https://github.com/pmem/pmemkv-python/blob/3989f2cc2ee41ed87eb1e510c4ee57952ecbe400/pmemkv/pmemkv.py#L139)):*
<details><summary><i>Previously, llugin wrote…</i></summary><blockquote>

1. not sure if part of this refactor but it's better if functions/classes documentations are docstrings, not comments. 

2a. if it is a public interface then the meaning of func should be explained. 2b. probably this could be an optional argument?
</blockquote></details>

It cannot be optional argument. Some brief documentation is added in this PR. Next step is to make it better.


<!-- Sent from Reviewable.io -->

igchor reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 6 files at r11, 1 of 1 files at r12.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20#-:-LyEjTQc9jJANUdhAaMm:bnfp4nl)* status: all files reviewed, 13 unresolved discussions (waiting on @llugin and @lukaszstolarczuk)



<!-- Sent from Reviewable.io -->

llugin reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/pmem/pmemkv-python/20#-:-LyEkkNfDWPfCdW9dFBf:bnfp4nl)* status: all files reviewed, 10 unresolved discussions (waiting on @lukaszstolarczuk)



<!-- Sent from Reviewable.io -->
